week4/problem.cpp

```diff
@@ -0,0 +1,61 @@
+class DiningPhilosophers {
+public:
+    typedef function<void()> fv;
+    int N = 5;
+    vector<mutex> locks = vector<mutex>(N);
+    vector<condition_variable> cvs = vector<condition_variable>(N);
+    vector<bool> taken = vector<bool>(N);
```

Comment:
vector<bool> is not thread safe, might store each bool as a bit, this might be the cause of the problem.
https://en.cppreference.com/w/cpp/container/vector_bool